docs/design/redundant-suppressions.md

```diff
@@ -0,0 +1,93 @@
+            }
+        }
+    }
+```
+
+Now the code is again trimmer-incompatible, the reflection pattern will trigger a `IL2075` warning. However, as we forgot to remove the suppression silencing the `IL2075` warnings, the warning will be suppressed. We will not be informed about this trimmer-incompatible pattern during trimming.
+
+
+## Detecting redundant warning suppressions
+
+In order to avoid the above scenario, we would like to have a option of detecting and reporting the warning suppressions which are not tied to any warnings caused trimmer-incompatible patterns.
+
+This may be achieved by extending the linker tool functionality to check which suppression do in fact suppress warnings and reporting those which do not.
+
+The said functionality can be implemented as an optional feature, triggered by passing a `
+--check-suppressions` command line argument to the linker. Running the tool with this option will report all of the warning suppressions which do not suppress any warnings.
+
+### Example:
+Let us again consider the example of the trimmer-compatible code with a redundant warning suppression. 
+
+```csharp
+    [UnconditionalSuppressMessage("trim", "IL2075", Justification = "It's OK to print only the properties which were actually used.")]
+    void PrintProperties<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties)] T>(T instance)
+    {
+        foreach (var p in typeof(T).GetProperties())
+        {
+            PrintPropertyValue(p, instance);
+        }
+    }
+
+```
+
+In order to detect the warning suppression not tied to any trimmer-incompatible pattern we should run the `dotnet publish` command and pass the `--check-suppressions` option.
+```shell
+  dotnet publish -r win-x64 -p:PublishTrimmed=True p:_ExtraTrimmerArgs="--check-suppressions"
+```
+
+The warning should be reported in the output of the command.
+
+```
+Trim analysis warning IL2021: Program.PrintProperties<T>(T): Unused UnconditionalSuppressMessageAttribute suppressing the IL2075 warning found. Consider removing the unused warning suppression.
+```
+
+## Other solutions
+
+The proposed solution operates by extending the functionality of the linker tool. On one hand, this allows for reusing the existing components leading to a simple implementation, on the other hand this may lead to potential problems. The linker sees only a part of code which is actually used, that means that the solution would not be able to identify the warning suppressions on the code which is trimmed away. Also, the linker may visit parts of the code which were not authored by the developer (e.g. used libraries) and look for the redundant suppressions there. This may cause the output warnings to be noisy and not useful to the developer. Moreover, the dependencies identified by the linker may be different depending on the environment the tool is run in. Hence, the proposed solution may report different redundant suppressions on different environments.
```

Comment:
Illink also does a bunch of optimizations that may render a suppression that was necessary at source compilation time look unnecessary. E.g. if constant propagation figures out a codepath is unreachable and suddenly the suppression looks unnecessary. but it's just unnecessary for that specific deployment mode.

The analyzer has a better view of the code from this perspective.